### PR TITLE
LWDEV-10241 Listen to changes in AssetPairsCache - notify disabledFuncationalityHolder when new assetPair added, add endpoint that returns last history records by date for all rules

### PR DESCRIPTION
**Pull request by rocketmannot**

### CODE REVIEW

File: src/main/kotlin/com/lykke/matching/engine/database/cache/AssetPairsCache.kt [hunk 1]
```diff
@@ -55,6 +64,13 @@ class AssetPairsCache @Autowired constructor (
             assetPairsById = newMap
             assetPairsByPair = newMapByPair
         }
+
+        val newAssetPairs = this.assetPairsById.values.filter { assetPair ->
+            knownAssetPairs.add(pairKey(assetPair.baseAssetId, assetPair.quotingAssetId))
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/database/cache/AssetPairsCache.kt]: Added new event for asset removal.

File: src/main/kotlin/com/lykke/matching/engine/holders/DisabledFunctionalityRulesHolder.kt [hunk 2]
```diff
@@ -136,6 +136,11 @@ class DisabledFunctionalityRulesHolder(val applicationSettingsCache: Application
         init()
     }
 
+    @EventListener
+    private fun onNewAssetPair(newAssetPairsEvent: NewAssetPairsEvent) {
+        init()
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/holders/DisabledFunctionalityRulesHolder.kt]: it is not needed - the field `disabledFunctionalityData` already has `@Volatile`

File: src/main/kotlin/com/lykke/matching/engine/database/cache/AssetPairsCache.kt [hunk 3]
```diff
@@ -2,28 +2,37 @@ package com.lykke.matching.engine.database.cache
...
+    @Volatile
     private var assetPairsById: Map<String, AssetPair> = HashMap()
+
+    @Volatile
     private var assetPairsByPair: Map<String, AssetPair> = HashMap()
 
     init {
         this.assetPairsById = databaseAccessor.loadAssetPairs()
         this.assetPairsByPair = generateAssetPairsMapByPair(assetPairsById)
+        assetPairsById.values.forEach { knownAssetPairs.add(pairKey(it.baseAssetId, it.quotingAssetId)) }
```
[the remark this answers is not in the document]
Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/database/cache/AssetPairsCache.kt]: changed

File: src/main/kotlin/com/lykke/matching/engine/database/cache/AssetPairsCache.kt [hunk 4]
```diff
@@ -59,18 +60,37 @@ class AssetPairsCache @Autowired constructor(
...
+
+        if (removedAssetPairs.isNotEmpty()) {
+            knownAssetPairs.removeAll(removedAssetPairs.map { it.assetPairId })
+            applicationEventPublisher.publishEvent(RemovedAssetPairsEvent(removedAssetPairs))
+        }
+    }
+
+    private fun getNewAssetPairs(newAssetPairs: Collection<AssetPair>): Collection<AssetPair> {
+        return newAssetPairs.filter { assetPair ->
+            knownAssetPairs.contains(assetPair.assetPairId)
```
Comment (barsukovskii) [inline src/main/kotlin/com/lykke/matching/engine/database/cache/AssetPairsCache.kt]: wrong filter, `knownAssetPairs.contains(assetPair.assetPairId)` -> `!knownAssetPairs.contains(assetPair.assetPairId)`

Comment (rocketmannot) [reply src/main/kotlin/com/lykke/matching/engine/database/cache/AssetPairsCache.kt]: fixed

**Pull request merged**